File: mobile/lib/data/models/goal_profile.dart
/// Goal-Adaptive Learning Layers — GoalProfile Dart twin.
///
/// Phase 6 of the Goal-Adaptive Learning Layers feature. Pure data + types.
/// Mirrors the contract in `src/lib/goals/goal-profile.ts` (web).
///
/// Authored: 2026-05-03
/// Owner: mobile (Dart twin) + assessment (rules)
///
/// Founder constraint: this file ships dormant. No mobile screen reads it
/// yet. The web-side feature flags (ff_goal_profiles, ff_goal_aware_foxy,
/// ff_goal_aware_selection, ff_goal_daily_plan, ff_goal_aware_rag,
/// ff_goal_daily_plan_reminder) are server-side and not consumed here;
/// the mobile UI surfaces this data lazily when the consumer widgets
/// are wired into the existing dashboard / parent / teacher screens.
///
/// Pure: zero Flutter imports, zero IO, zero side effects. All strings
/// are author-written literals. Web ↔ mobile string equivalence is
/// pinned by the test suite (test/data/models/goal_profile_test.dart).

library;

/// All 6 academic goals a student can choose. Strings match the
/// web-side TypeScript GoalCode union exactly.
enum GoalCode {
  improveBasics('improve_basics'),
  passComfortably('pass_comfortably'),
  schoolTopper('school_topper'),
  boardTopper('board_topper'),
  competitiveExam('competitive_exam'),
  olympiad('olympiad');

  final String code;
  const GoalCode(this.code);

  /// Resolve a raw string code (e.g. 'board_topper') to the enum value,
  /// or return null for unknown / null / empty input.
  static GoalCode? fromCode(String? code) {
    if (code == null || code.isEmpty) return null;
    for (final g in GoalCode.values) {
      if (g.code == code) return g;
    }
    return null;
  }
}

/// Difficulty distribution: easy + medium + hard sums to 1.0 (± 1e-9).
class DifficultyMix {
  final double easy;
  final double medium;
  final double hard;
  const DifficultyMix({required this.easy, required this.medium, required this.hard});
}

/// Bloom's taxonomy band: 1=remember, 2=understand, 3=apply, 4=analyze,
/// 5=evaluate, 6=create. min ≤ max.
class BloomBand {
  final int min;
  final int max;
  const BloomBand({required this.min, required this.max});
}

enum SourceTag {
  ncert('ncert'),
  pyq('pyq'),
  jeeArchive('jee_archive'),
  neetArchive('neet_archive'),
  olympiad('olympiad'),
  curated('curated');

  final String tag;
  const SourceTag(this.tag);
}

enum PacePolicy {
  patient('patient'),
  steady('steady'),
  push('push'),
  campaign('campaign'),
  selective('selective');

  final String value;
  const PacePolicy(this.value);
}

enum ScorecardTone {
  encouraging('encouraging'),
  analytical('analytical'),
  examiner('examiner');

  final String value;
  const ScorecardTone(this.value);
}

class GoalProfile {
  final GoalCode code;
  final String labelEn;
  final String labelHi;
  final DifficultyMix difficultyMix;
  final BloomBand bloomBand;
  final List<SourceTag> sourcePriority;
  final double masteryThreshold;
  final int dailyTargetMinutes;
  final PacePolicy pacePolicy;
  final ScorecardTone scorecardTone;
  final String dashboardCalloutEn;
  final String dashboardCalloutHi;

  const GoalProfile({
    required this.code,
    required this.labelEn,
    required this.labelHi,
    required this.difficultyMix,
    required this.bloomBand,
    required this.sourcePriority,
    required this.masteryThreshold,
    required this.dailyTargetMinutes,
    required this.pacePolicy,
    required this.scorecardTone,
    required this.dashboardCalloutEn,
    required this.dashboardCalloutHi,
  });
}

/// All 6 GoalProfiles. Values MUST stay in lockstep with
/// src/lib/goals/goal-profile.ts on the web side. Drift is detected
/// by integration tests that pull both sides.
const Map<GoalCode, GoalProfile> goalProfiles = {
  GoalCode.improveBasics: GoalProfile(
    code: GoalCode.improveBasics,
    labelEn: 'Improve Basics',
    labelHi: 'बेसिक्स सुधारें',
    difficultyMix: DifficultyMix(easy: 0.60, medium: 0.35, hard: 0.05),
    bloomBand: BloomBand(min: 1, max: 3),
    sourcePriority: [SourceTag.ncert, SourceTag.curated],
    masteryThreshold: 0.60,
    dailyTargetMinutes: 10,
    pacePolicy: PacePolicy.patient,
    scorecardTone: ScorecardTone.encouraging,
    dashboardCalloutEn: 'Today: one concept at a time',
    dashboardCalloutHi: 'आज: एक-एक अवधारणा पर ध्यान',
  ),
  GoalCode.passComfortably: GoalProfile(
    code: GoalCode.passComfortably,
    labelEn: 'Pass Comfortably',
    labelHi: 'आराम से पास',
    difficultyMix: DifficultyMix(easy: 0.40, medium: 0.45, hard: 0.15),
    bloomBand: BloomBand(min: 1, max: 4),
    sourcePriority: [SourceTag.ncert, SourceTag.pyq],
    masteryThreshold: 0.70,
    dailyTargetMinutes: 20,
    pacePolicy: PacePolicy.steady,
    scorecardTone: ScorecardTone.encouraging,
    dashboardCalloutEn: 'Top board-frequency topics today',
    dashboardCalloutHi: 'बोर्ड में सबसे अधिक पूछे जाने वाले विषय आज',
  ),
  GoalCode.schoolTopper: GoalProfile(
    code: GoalCode.schoolTopper,
    labelEn: 'School Topper',
    labelHi: 'स्कूल टॉपर',
    difficultyMix: DifficultyMix(easy: 0.30, medium: 0.50, hard: 0.20),
    bloomBand: BloomBand(min: 1, max: 5),
    sourcePriority: [SourceTag.ncert, SourceTag.pyq, SourceTag.curated],
    masteryThreshold: 0.80,
    dailyTargetMinutes: 30,
    pacePolicy: PacePolicy.push,
    scorecardTone: ScorecardTone.analytical,
    dashboardCalloutEn: 'Today: depth + application',
    dashboardCalloutHi: 'आज: गहराई और अनुप्रयोग',
  ),
  GoalCode.boardTopper: GoalProfile(
    code: GoalCode.boardTopper,
    labelEn: 'Board Topper (90%+)',
    labelHi: 'बोर्ड टॉपर (90%+)',
    difficultyMix: DifficultyMix(easy: 0.20, medium: 0.45, hard: 0.35),
    bloomBand: BloomBand(min: 2, max: 6),
    sourcePriority: [SourceTag.pyq, SourceTag.ncert, SourceTag.curated],
    masteryThreshold: 0.85,
    dailyTargetMinutes: 45,
    pacePolicy: PacePolicy.campaign,
    scorecardTone: ScorecardTone.examiner,
    dashboardCalloutEn: "Today's PYQ streak: 5 board questions",
    dashboardCalloutHi: 'आज की PYQ श्रृंखला: 5 बोर्ड प्रश्न',
  ),
  GoalCode.competitiveExam: GoalProfile(
    code: GoalCode.competitiveExam,
    labelEn: 'JEE/NEET Prep',
    labelHi: 'JEE/NEET तैयारी',
    difficultyMix: DifficultyMix(easy: 0.10, medium: 0.40, hard: 0.50),
    bloomBand: BloomBand(min: 3, max: 6),
    sourcePriority: [SourceTag.jeeArchive, SourceTag.neetArchive, SourceTag.ncert, SourceTag.curated],
    masteryThreshold: 0.85,
    dailyTargetMinutes: 60,
    pacePolicy: PacePolicy.campaign,
    scorecardTone: ScorecardTone.analytical,
    dashboardCalloutEn: 'JEE/NEET targeted set today',
    dashboardCalloutHi: 'आज JEE/NEET लक्षित अभ्यास',
  ),
  GoalCode.olympiad: GoalProfile(
    code: GoalCode.olympiad,
    labelEn: 'Olympiad Prep',
    labelHi: 'ओलंपियाड',
    difficultyMix: DifficultyMix(easy: 0.05, medium: 0.25, hard: 0.70),
    bloomBand: BloomBand(min: 4, max: 6),
    sourcePriority: [SourceTag.olympiad, SourceTag.ncert, SourceTag.curated],
    masteryThreshold: 0.90,
    dailyTargetMinutes: 60,
    pacePolicy: PacePolicy.selective,
    scorecardTone: ScorecardTone.analytical,
    dashboardCalloutEn: '1 olympiad challenge today',
    dashboardCalloutHi: 'आज 1 ओलंपियाड चुनौती',
  ),
};

/// Resolve a goal profile from a raw string code. Returns null for
/// null / empty / unknown values.
GoalProfile? resolveGoalProfile(String? code) {
  final goal = GoalCode.fromCode(code);
  if (goal == null) return null;
  return goalProfiles[goal];
}

File: packages/lib/src/goals/goal-profile.ts
/**
 * Alfanumrik — Goal-Adaptive Learning Layers / Phase 0
 * Goal Profile Resolver
 *
 * Owner: assessment
 * Founder constraint: this file MUST NOT modify any existing file. Pure new module.
 * Other agents will wire callers in behind feature flags.
 *
 * Single source of truth for the academic-goal persona table. Every downstream
 * consumer (Foxy persona, scorecard sentence, quiz selection mix, study-plan
 * pacing, progress dashboard callouts) MUST read from `GOAL_PROFILES` here
 * rather than hardcoding values per surface.
 *
 * Consumers (will read this in later wiring PRs):
 *  - src/lib/ai/prompts/foxy-system.ts  (replace single-line GOAL_PROMPT_MAP entry
 *                                        with buildExpandedGoalSection(...) when
 *                                        ff_goal_aware_foxy is on)
 *  - src/lib/goals/goal-personas.ts     (Phase 1 expanded persona — same dir)
 *  - src/lib/goals/scorecard-sentence.ts (Phase 1 scorecard tone — same dir)
 *  - src/components/quiz/QuizResults.tsx (scorecard sentence display — Phase 1)
 *  - src/lib/cognitive-engine.ts        (Phase 2 — difficultyMix + bloomBand
 *                                        weighting in question selection)
 *  - supabase/functions/quiz-generator/  (Phase 2 — sourcePriority filter)
 *
 * Feature flags that gate this module's effect:
 *  - ff_goal_profiles      (master switch — when off, callers fall back to legacy)
 *  - ff_goal_aware_foxy    (Phase 1 — Foxy persona uses buildExpandedGoalSection)
 *  - ff_goal_aware_scoring (Phase 1 — QuizResults shows goal-aware sentence)
 *
 * Pure data + types. ZERO IO, ZERO React, ZERO side effects, ZERO PII handling.
 * All strings are author-written literals — no LLM-generated text in this file.
 *
 * Invariants:
 *  - difficultyMix.{easy,medium,hard} sums to 1.0 within ±1e-9
 *  - bloomBand.min ≤ bloomBand.max, both in {1..6}
 *  - masteryThreshold ∈ (0, 1]
 *  - dailyTargetMinutes > 0
 *  - resolveGoalProfile NEVER throws — returns null on any unknown input
 *  - GOAL_PROFILES is deeply Readonly (TS) and frozen at runtime
 *
 * P5: grade-format invariant is unaffected — this module doesn't carry grade values.
 * P7: bilingual labels and callouts authored in pairs (en + hi).
 */

// ─── Types ────────────────────────────────────────────────────────────────

export type GoalCode =
  | 'board_topper'
  | 'school_topper'
  | 'pass_comfortably'
  | 'competitive_exam'
  | 'olympiad'
  | 'improve_basics';

/** Probabilities sum to 1.0 (within ±1e-9). */
export type DifficultyMix = {
  easy: number;
  medium: number;
  hard: number;
};

/** Bloom's taxonomy levels: 1=remember, 2=understand, 3=apply, 4=analyze,
 *  5=evaluate, 6=create. min ≤ max. */
export type BloomBand = {
  min: 1 | 2 | 3 | 4 | 5 | 6;
  max: 1 | 2 | 3 | 4 | 5 | 6;
};

/** Question-source tags consulted in priority order (first = highest priority). */
export type SourceTag =
  | 'ncert'
  | 'pyq'
  | 'jee_archive'
  | 'neet_archive'
  | 'olympiad'
  | 'curated';

/** Pacing policy for study plans + Foxy mode-selection bias. */
export type PacePolicy =
  | 'patient'
  | 'steady'
  | 'push'
  | 'campaign'
  | 'selective';

/** Tone applied to scorecard sentences after a quiz attempt. */
export type ScorecardTone = 'encouraging' | 'analytical' | 'examiner';

export interface GoalProfile {
  code: GoalCode;
  labelEn: string;
  labelHi: string;
  difficultyMix: DifficultyMix;
  bloomBand: BloomBand;
  sourcePriority: SourceTag[];
  /** Mastery cutoff for "concept mastered" decisions, 0-1. */
  masteryThreshold: number;
  /** Daily study-time recommendation surfaced by progress + study-plan UIs. */
  dailyTargetMinutes: number;
  pacePolicy: PacePolicy;
  scorecardTone: ScorecardTone;
  /** Short EN sentence shown in the dashboard goal callout strip. */
  dashboardCalloutEn: string;
  /** Short HI sentence shown in the dashboard goal callout strip. */
  dashboardCalloutHi: string;
}

// ─── Profile table ────────────────────────────────────────────────────────
//
// Author-written. Edit ONLY with assessment-agent sign-off. Mirrors the
// founder-approved persona table for the Goal-Adaptive Learning Layers spec.

const PROFILES: Record<GoalCode, GoalProfile> = {
  improve_basics: {
    code: 'improve_basics',
    labelEn: 'Improve Basics',
    labelHi: 'बेसिक्स सुधारें',
    difficultyMix: { easy: 0.6, medium: 0.35, hard: 0.05 },
    bloomBand: { min: 1, max: 3 },
    sourcePriority: ['ncert', 'curated'],
    masteryThreshold: 0.6,
    dailyTargetMinutes: 10,
    pacePolicy: 'patient',
    scorecardTone: 'encouraging',
    dashboardCalloutEn: 'Small steps today: 10 minutes is enough to build your base.',
    dashboardCalloutHi: 'आज छोटे कदम: बेसिक्स पक्के करने के लिए 10 मिनट काफी हैं।',
  },
  pass_comfortably: {
    code: 'pass_comfortably',
    labelEn: 'Pass Comfortably',
    labelHi: 'आराम से पास',
    difficultyMix: { easy: 0.4, medium: 0.45, hard: 0.15 },
    bloomBand: { min: 1, max: 4 },
    sourcePriority: ['ncert', 'pyq'],
    masteryThreshold: 0.7,
    dailyTargetMinutes: 20,
    pacePolicy: 'steady',
    scorecardTone: 'encouraging',
    dashboardCalloutEn: 'Steady wins boards: 20 focused minutes on high-frequency topics today.',
    dashboardCalloutHi: 'धीरे-धीरे बोर्ड पक्का: आज 20 मिनट ज़रूरी टॉपिक्स पर लगाओ।',
  },
  school_topper: {
    code: 'school_topper',
    labelEn: 'School Topper',
    labelHi: 'स्कूल टॉपर',
    difficultyMix: { easy: 0.3, medium: 0.5, hard: 0.2 },
    bloomBand: { min: 1, max: 5 },
    sourcePriority: ['ncert', 'pyq', 'curated'],
    masteryThreshold: 0.8,
    dailyTargetMinutes: 30,
    pacePolicy: 'push',
    scorecardTone: 'analytical',
    dashboardCalloutEn: '30 minutes today — push past textbook into application questions.',
    dashboardCalloutHi: 'आज 30 मिनट — किताब से आगे एप्लीकेशन सवालों पर जाओ।',
  },
  board_topper: {
    code: 'board_topper',
    labelEn: 'Board Topper (90%+)',
    labelHi: 'बोर्ड टॉपर (90%+)',
    difficultyMix: { easy: 0.2, medium: 0.45, hard: 0.35 },
    bloomBand: { min: 2, max: 6 },
    sourcePriority: ['pyq', 'ncert', 'curated'],
    masteryThreshold: 0.85,
    dailyTargetMinutes: 45,
    pacePolicy: 'campaign',
    scorecardTone: 'examiner',
    dashboardCalloutEn: "Today's PYQ streak: 5 board questions — examiner mindset on.",
    dashboardCalloutHi: 'आज की PYQ स्ट्रीक: 5 बोर्ड सवाल — एग्ज़ामिनर माइंडसेट।',
  },
  competitive_exam: {
    code: 'competitive_exam',
    labelEn: 'JEE/NEET Preparation',
    labelHi: 'JEE/NEET तैयारी',
    difficultyMix: { easy: 0.1, medium: 0.4, hard: 0.5 },
    bloomBand: { min: 3, max: 6 },
    sourcePriority: ['jee_archive', 'neet_archive', 'ncert', 'curated'],
    masteryThreshold: 0.85,
    dailyTargetMinutes: 60,
    pacePolicy: 'campaign',
    scorecardTone: 'analytical',
    dashboardCalloutEn: '60 minutes today — timed JEE/NEET archive set, watch your pace.',
    dashboardCalloutHi: 'आज 60 मिनट — JEE/NEET आर्काइव सेट, रफ़्तार पर ध्यान दो।',
  },
  olympiad: {
    code: 'olympiad',
    labelEn: 'Olympiad',
    labelHi: 'ओलंपियाड',
    difficultyMix: { easy: 0.05, medium: 0.25, hard: 0.7 },
    bloomBand: { min: 4, max: 6 },
    sourcePriority: ['olympiad', 'ncert', 'curated'],
    masteryThreshold: 0.9,
    dailyTargetMinutes: 60,
    pacePolicy: 'selective',
    scorecardTone: 'analytical',
    dashboardCalloutEn: 'One olympiad-grade problem today — multiple solution paths welcome.',
    dashboardCalloutHi: 'आज एक ओलंपियाड-स्तर का सवाल — कई हल अप्रोच आज़माओ।',
  },
};

// Deep-freeze: Readonly type at compile time, Object.freeze at runtime so any
// accidental mutation throws in dev/test (catches bugs in callers).
function deepFreeze<T>(obj: T): Readonly<T> {
  if (obj && typeof obj === 'object') {
    for (const key of Object.keys(obj as Record<string, unknown>)) {
      const value = (obj as Record<string, unknown>)[key];
      if (value && typeof value === 'object') {
        deepFreeze(value);
      }
    }
    Object.freeze(obj);
  }
  return obj as Readonly<T>;
}

export const GOAL_PROFILES: Readonly<Record<GoalCode, GoalProfile>> =
  deepFreeze(PROFILES);

// ─── Helpers ──────────────────────────────────────────────────────────────

const KNOWN_GOAL_CODES = new Set<GoalCode>([
  'board_topper',
  'school_topper',
  'pass_comfortably',
  'competitive_exam',
  'olympiad',
  'improve_basics',
]);

/**
 * True if `value` is a recognized GoalCode string. Discriminating type guard.
 * Safe to call on any input — never throws.
 */
export function isKnownGoalCode(value: unknown): value is GoalCode {
  return typeof value === 'string' && KNOWN_GOAL_CODES.has(value as GoalCode);
}

/**
 * Resolve a stored `students.academic_goal` value to its frozen GoalProfile.
 * Returns null for null/undefined/empty/unknown — callers MUST handle null
 * by falling back to legacy behavior (single-line GOAL_PROMPT_MAP, default
 * scorecard line, etc). Never throws.
 */
export function resolveGoalProfile(
  code: string | null | undefined,
): GoalProfile | null {
  if (!code) return null;
  if (!isKnownGoalCode(code)) return null;
  return GOAL_PROFILES[code];
}
